draft-thomson-train-protocol.md

```diff
@@ -99,167 +105,322 @@ cannot assume that rate limit indications from one path apply to new paths.
 
 
-# Version Mappings
-
-This document defines version number mappings for QUIC version 1 {{!RFC9000}}
-and version 2 {{!RFC9369}}.  Rate limit mappings are defined for a limited
-number of send rates, as shown in {{table-rates}}.
+# TRAIN Packet {#packet}
 
-{:aside}
-> Note: The exact set of rates that are included is subject to negotiation.
+A TRAIN packet is a QUIC long header packet that follows the QUIC invariants;
+see {{Section 5.1 of INVARIANTS}}.
 
-| Rate Limit | Version 1 | Version 2 |
-|--:|:--|:--|
-| 1Mbps      | 0xTBD | 0xTBD |
-| 2Mbps      | 0xTBD | 0xTBD |
-| 3Mbps      | 0xTBD | 0xTBD |
-| 4Mbps      | 0xTBD | 0xTBD |
-| 5Mbps      | 0xTBD | 0xTBD |
-| 7Mbps      | 0xTBD | 0xTBD |
-| 10Mbps      | 0xTBD | 0xTBD |
-| 15Mbps      | 0xTBD | 0xTBD |
-| 20Mbps      | 0xTBD | 0xTBD |
-| 50Mbps      | 0xTBD | 0xTBD |
-{: #table-rates title="Version and Rate Limit Mappings"}
-
-Packets for each of these versions is processed exactly as they would be for
-QUIC version 1 or QUIC version 2.  In particular, when a packet with one of
-these versions is received, the value of the version field is replaced by the
-version number for either QUIC version 1 (0x00000001) or QUIC version 2
-(0x6b3343cf).  Thus, the packet protection mechanism used by the canonical QUIC
-version is identical and AEAD decryption would fail if the limit-specific
-version number is left in place.
-
-For example, the example Retry packet from {{Appendix A.4 of RFC9369}} might be
-modified to indicate a rate limit of TBD to produce a packet of:
+{{fig-train-packet}} shows the format of the train packet using the conventions
+from {{Section 4 of INVARIANTS}}.
 
+~~~ artwork
+TRAIN Packet {
+  Header Form (1) = 1,
+  Reserved (1),
+  Rate Signal (6),
+  Version (32) = 0xTBD,
+  Destination Connection ID Length (8),
+  Destination Connection ID (0..2040),
+  Source Connection ID Length (8) = 0,
+}
 ~~~
-cfXXXXXXXX0008f067a5502a4262b574 6f6b656ec8646ce8bfe33952d9555436
-65dcc7b6
-~~~
+{: #fig-train-packet title="TRAIN Packet Format"}
 
-But the 4 bytes of the version number will be restored to 0x6b3343cf before
-attempting to process the packet further.
+The most significant bit (0x80) of the packet indicates that this is a QUIC long
+header packet.  The next bit (0x40) is reserved and can be set according to
+{{!QUIC-BIT=RFC9287}}.
 
+The entire payload of the TRAIN packet is carried in the Rate Signal field that
+forms the low 6 bits (0x3f) of the first byte.  Values for this field are
+described in {{rate-signal}}.
 
-# Deployment
+This packet includes a Destination Connection ID field that is set to the same
+value as other packets in the same datagram; see {{Section 12.2 of QUIC}}.
+
+The Source Connection ID Length field is set to zero, meaning that the Source
+Connection ID field is empty.  TRAIN packets can still be sent during the QUIC
+handshake; the Source Connection ID field in other packets might be needed to
+successfully complete a handshake.
```

Comment:
I wonder if this is a needless restriction due to following reasons:
* Future versions of QUIC might use SCID for different purposes.
* QUIC v1 allows use of long header packets mid-connection (see [Section 14.4.1 of RFC 9000](https://datatracker.ietf.org/doc/html/rfc9000#name-pmtu-probes-containing-sour)), but there is no such restriction. I'm not sure if there is a reason to diverge from the design that we already have.

My preference of moving forward would be to state SCID and DCID fields should be filled as defined by the QUIC version used by the endpoints, and add a reference to the section of RFC 9000 pointed out above.

Reply:
Thanks.

Am I correct to interpret what you are saying is that, for train packets used by QUIC v1, we can define how the SCID field is filled, but we might want to keep room for future QUIC versions to do something else?

If that's your argument, you have my +1.

Reply:
I think the format of the TRAIN packet must be defined by the TRAIN version number. The parser logic is "look at the VN, recognize the version and the packet type, parse accordingly." This does not involve looking at other coalesced packets in the same datagram. If we decide to change the format, we will have to set VN=TRAIN2.

Reply:
RFC 9000 section 12.2 says: "Receivers MAY route based on the information in the first packet contained in a UDP datagram. Senders MUST NOT coalesce QUIC packets with different connection IDs into a single UDP datagram. Receivers SHOULD ignore any subsequent packets with a different Destination Connection ID than the first packet in the datagram.

Suppose a coalesced packet with: "Long, V=TRAIN, DCID=X, SCID=NULL", then "Long, type=handshake, V= 1, DCID=X, SCID=S". Arguably, it does not meet the "MUST NOT", because the SCID of the first and second packets have different SCID.

If we want to not break that RFC 9000 rule, then if TRAIN packets are coalesced with long header packets, the SCID MUST be set to the same value used in the long header packets.

Reply:
On the other hand, if the TRAIN packet is coalesced only with short header packet, then filling in the SCID is arguably a privacy leak.

Reply:
Let's merge this PR and continue the discussion on #13.